### ISSUE TITLE
Performance Issue in nmk_kernel

### ISSUE THREAD

**Issue by FindHao** [pytorch/vision#6403]
### 🚀 The feature

Hi, when analyzing the profiling results of [detectron2_fasterrcnn_r_101_fpn](https://github.com/pytorch/benchmark) in inference mode,  I found some inefficiencies. More details described in the following box.

### Motivation, pitch

The batch size is 4, so there will be 4 calls for `batched_nms`. Deep to the bottom of the calling context tree, there are some long calls for`cudaMemcpyAsync` and following empty.
<img width="1152" alt="Screen Shot 2022-08-11 at 4 42 01 PM" src="https://user-images.githubusercontent.com/3823240/184260161-5767aa3b-cf62-40e6-a097-976078d5778a.png">

The corresponding source code is the following part. Line 136 is the memory copy, and I believe the following empty are the for loop at Line 148-Line 159. 

https://github.com/pytorch/vision/blob/7fb8d068618f6b913f3d8489b74bf95911184e0f/torchvision/csrc/ops/cuda/nms_kernel.cu#L136-L164

From the profiling results I collected, if we can make it more efficient like half of the current execution time for those lines, the speedup for detectron2 series model would be 1.02X and for timm_efficientdet would be 1.15X. 

I did try to change it on GPU. please see [this minimal reproduce repo](https://github.com/FindHao/test_nms_kernel/blob/master/test.cu). But it looks simply moving the current logic to GPU will make it worse because of the data dependency. In my testing, the GPU version is 1.40X slower than CPU version. But I'll leave the repo there to inspire more people and make the testing simple.  

I'm not so familiar with the algorithm here, is there anyway to get rid of the data dependency here?

**Comment by datumbox** [comment]
@FindHao thanks for documenting this.

To ensure I understand you right, you are saying that the copy from CPU to GPU of line 136 seemed to you suspicious on the profiler but once you tried to move it to avoid it (by keeping the mask on GPU) you get a 1.4x slowdown. Is this an accurate depiction of the situation?

I am not the author of NMS either. I believe @fmassa wrote it. He could potentially provide some additional context from his side over the choices of the implementation.

**Comment by FindHao** [comment]
@datumbox yes. you are correct. To avoid it, not only keeping the mask on GPU, but also moving the following for loop to GPU. But there are some data dependencies between iterations in the for loop and my implementation is quite naive, so there are 1.4X slowdown(when the input is smaller, it could be 2-10X slowdown for this part).

**Comment by NicolasHug** [comment]
I think this should be addressed by https://github.com/pytorch/vision/pull/8766
